Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,346 @@
+        [self->_testCase assertNilErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
-}
 
-- (void)testDeleteDocument {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
-  [self writeDocumentRef:doc data:@{@"foo" : @"bar"}];
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
-  XCTAssertEqualObjects(@"bar", snapshot[@"foo"]);
+  [_testCase awaitExpectations];
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
+- (void)runFailingTransaction:(FIRFirestoreErrorCode)expected {
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"runTransaction"];
+  [_db
       runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        [transaction deleteDocument:doc];
+        for (TransactionStage stage in self->_stages) {
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertEqualObjects(@YES, result);
-        XCTAssertNil(error);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence (%@), to fail, but it didn't.",
+                                       [self listStages]];
+        [self->_testCase assertErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
 
-  snapshot = [self readDocumentForRef:doc];
+  [_testCase awaitExpectations];
+}
+
+- (void)cleanupTester {
+  _stages = [NSArray array];
+  // Set the docRef to something else to lose the original reference.
+  _docRef = [[self->_db collectionWithPath:@"reset"] documentWithAutoID];
+}
+
+- (NSString *)listStages {
+  NSMutableArray<NSString *> *seqList = [NSMutableArray array];
+  for (TransactionStage stage in _stages) {
+    if (stage == delete1) {
+      [seqList addObject:@"delete"];
+    } else if (stage == update1 || stage == update2) {
+      [seqList addObject:@"update"];
+    } else if (stage == set1 || stage == set2) {
+      [seqList addObject:@"set"];
+    } else if (stage == get) {
+      [seqList addObject:@"get"];
+    }
+  }
+  return [seqList description];
+}
+
+@end
+
+@implementation FSTTransactionTests
+- (void)assertExistsWithSnapshot:(FIRDocumentSnapshot *)snapshot error:(NSError *)error {
+  XCTAssertNil(error);
+  XCTAssertTrue(snapshot.exists);
+}
+
+- (void)assertDoesNotExistWithSnapshot:(FIRDocumentSnapshot *)snapshot error:(NSError *)error {
+  XCTAssertNil(error);
   XCTAssertFalse(snapshot.exists);
 }
 
-- (void)testGetNonexistentDocumentThenCreate {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
+- (void)assertNilErrorWithMessage:(NSString *)message error:(NSError *)error {
```

Comment:
This method reverses the arguments of the assertion below. Consider an alternative that preserves the ordering. Maybe `assertNilError:message:`?

Similarly `assertErrorWithMessage:error:` reverses the arguments.

Note that `NSError` is normally last when it's being used as an out parameter, a callback error parameter, or otherwise involved in actual error handling. Here it's the primary object of the assertion so it should be first.